src/systems/ground_reveal.rs

```diff
@@ -0,0 +1,91 @@
+use amethyst::core::specs::prelude::{Read, ReadStorage, System, Write, WriteStorage};
+use amethyst::core::timing::Time;
+use amethyst::core::transform::Transform;
+use amethyst::ecs::Entity;
+
+use entities::tile::LevelGrid;
+use entities::tile::Tile;
+use std::cmp::Reverse;
+use std::collections::BinaryHeap;
+use std::time::Duration;
+use systems::TileUpdateQueue;
+
+///
+/// This system handles the recursive process of revealing tiles
+/// including telling the adjacent tiles to update their meshes
+///
+pub struct GroundRevealSystem;
+
+impl<'a> System<'a> for GroundRevealSystem {
+    type SystemData = (
+        Read<'a, Time>,
+        Read<'a, LevelGrid>,
+        Write<'a, BinaryHeap<Reverse<(Duration, Entity)>>>,
+        ReadStorage<'a, Transform>,
+        WriteStorage<'a, Tile>,
+        Write<'a, TileUpdateQueue>,
+    );
+
+    fn run(
+        &mut self,
+        (time, grid, mut heap, transforms, mut tiles, mut tile_update_queue): Self::SystemData,
+    ) {
+        if let Some(Reverse((mut reveal_time, mut entity))) = heap.peek().cloned() {
+            while reveal_time <= time.absolute_time() {
+                //the entity is to be revealed, so we delete it, but we already got the values by peeking
+                heap.pop();
+
+                let tran = transforms.get(entity).unwrap().clone();
+                let x = tran.translation[0] as i32;
+                let y = -tran.translation[2] as i32;
+
+                // reveal yourself
+                tiles.get_mut(entity).unwrap().reveal();
+
+                let mut neighbors = vec![];
+                neighbors.extend(grid.direct_neighbors(x, y));
+
+                for neighbor in neighbors.clone().iter() {
+                    // add conceiled to queue
+                    let tile = tiles.get_mut(*neighbor).unwrap();
+                    match tile {
+                        Tile::Ground { concealed: true } => {
+                            heap.push(Reverse((
+                                Duration::from_millis(200) + time.absolute_time(),
+                                neighbor.clone(),
+                            )));
+
+                            let pos = neighbors.iter().position(|x| *x == *neighbor).unwrap();
+                            neighbors.remove(pos);
```

Comment:
you're cloning both the whole container and then every individual element, and then iterating through the original container to find the position to remove that element! there *has* to be another way.

Reply:
true